### ISSUE TITLE
[Feature Request] prune the optuna trial

### ISSUE THREAD

**Issue by chris-tkinter** [facebookresearch/hydra#1710]
# 🚀 Feature Request

Hi, I know that we can use the optuna plugin to maximize (or minimize) the return result of the function wrapped by the `@hydra.main` decorator. 

However, I would also like to prune the experiment if optuna thinks the experiment would be fruitless. In plain pytorch it can be done by [this](https://github.com/optuna/optuna-examples/blob/main/pytorch/pytorch_simple.py#L118-L119)

I am wondering if there is a workaround for this, as we know have no access to the `trail` parameter I think.

Thanks!

(As a little more context, I am trying to minimize the validation loss and would like to have an "early stop" if the validation loss isn't decreasing.)

**Comment by omry** [comment]
It's not clear what is the feature request here. Do you want the launcher to print the number of pruned trials?

**Comment by chris-tkinter** [comment]
Hi, sorry for the confusion, I would like to have access to the optuna's `trail` variable (the variable in [here](https://github.com/optuna/optuna-examples/blob/main/pytorch/pytorch_simple.py#L118-L119)) when using the optuna sweeper so that I can prune it. Is it possible?

**Comment by omry** [comment]
How are you proposing to have access to it exactly? (That object is hidden inside the state of the sweeper).

**Comment by chris-tkinter** [comment]
I was thinking about passing the sweeper instance as an additional optional parameter to the function decorated by the `hydra.main`. Although I haven't taken a close look at the current implementation but I think this is a non-trivial change

**Comment by omry** [comment]
This will break the layering in the architecture and is not something I would consider.

What do you need to do with the pruning results?

**Comment by chris-tkinter** [comment]
I see. Thanks. Closing this issue now.

**Comment by omry** [comment]
Something that might work is to add support for pruning to the Optuna sweeper itself.
If you are interested in sending a pull request for that let me know. we can discuss what it might look like before you spend time on it.

**Comment by robogast** [comment]
I'm looking at adding a pruner to the `optuna` plugin, but I don't see how adding a pruner is possible without explicitly passing a `study` or `trial` object through the `launcher`:

In their short [tutorial](https://optuna.readthedocs.io/en/stable/tutorial/10_key_features/004_distributed.html) on optuna+multiprocessing, they explicitly load the `study` object in the independent process, which reads a database and carries on from there.

As an example, I would like to use the [PyTorch Lightning pruning callback](https://optuna.readthedocs.io/en/stable/_modules/optuna/integration/pytorch_lightning.html#PyTorchLightningPruningCallback), but I don't see how I can pass/load the trial in any meaningful way.

I'm willing to help with a PR, but I could use some pointers :).

**Comment by Jasha10** [comment]
Do I understand correctly that this pruning would require communication between the running job and the sweeper (before the job has finished)?

This may be an area where experimentation is needed (with the goal of generalizing across different launchers / across different remote execution environments).

**Comment by MartinSmeyer** [comment]
I also think the combination of hydra with optuna+submitit plugins would be awesome but it lacks the pruning feature (and also the loading/saving of existing studies which is easier to add, though). 

@Jasha10 Yes, the job needs to communicate intermediate results back to the sweeper and this should work among different launchers/remotes. In my case, I did not see a way around saving intermediate result files tediously in our shared filesystem and loading them directly in the sweeper. 

Also the launchers currently are synchronous, so they wait until all results have finished - meaning pruning in batch mode would not really help. The problem seems to be recognized (nevergrad sweeper):

https://github.com/facebookresearch/hydra/blob/30a8c8a070166ac2d3b19c4ca29af7aa500dc269/plugins/hydra_nevergrad_sweeper/hydra_plugins/hydra_nevergrad_sweeper/_impl.py#L162-L164

**Comment by robogast** [comment]
I've forked the repo, and modified some of the optuna plugin code so my implementation now supports:
- asynchronous launching/returning (at least with slurm, using joblib.Parallel)
- pruners (by passing the trial object to each individual job)

## Paralellism (with slurm submitit backend)
Actually relatively simple logic using the `joblib` library, simply wrap the trial ask/tell logic in a function, and delegate paralellism to joblibs `Parallel`  

https://github.com/sara-nl/hydra/blob/050b90b4a8c3385c67c98ad3c8a64711108df034/plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py#L216-L224

The ask/tell logic is wrapped here:

https://github.com/sara-nl/hydra/blob/050b90b4a8c3385c67c98ad3c8a64711108df034/plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py#L253

The downside to this method is that the launcher no longer submits array jobs, but submits jobs in a for loop (but since you practically submit only a handful of jobs at a time, this is not an issue, at least for me).
 
## Pruning
- Make sure the study uses persistent storage (I pass an sqlite db path in the optuna config) 
- Add the pruner as an option to the config (for now I only added the hyperband pruner)
- Pass the pruner to the study
- Dump the pickled trial to a tempfile, for which I create an instantiable config, instantiate it in the train script, and call `trial.should_prune()` in the training script (I use the `PyTorchLightningPruningCallback`, as mentioned before)
https://github.com/sara-nl/2D-VQ-AE-2/blob/6708eeafdbdbb0c79f4a942b378206f6b1ba2580/vq_ae/train.py#L18-L21

The code is definitely not production-ready (for one, I now leak a file-descriptor when instantiating the trial object), but I'm curious to what you think about it :)
